File: red/src/describe.ts
import * as machine from "./machine.ts";
import { existsSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { runtime, type ExecResult } from "red/runtime";
import type { Opts } from "red/workflow";
import { readPars } from "red/cli";
import { identityArgs, withMachineKey } from "./ssh.ts";
import { backendCredentialEnv, toolDir } from "./tools.ts";

const runTimeoutMs = 30_000;
const sshProbeTimeoutMs = 10_000;
const registryTimeoutMs = 30_000;
const placeholderIp = "192.168.0.1";

type Runner = (args: string[], opts?: { cwd?: string; env?: Record<string, string | undefined>; timeoutMs?: number }) => Promise<ExecResult>;
const run: Runner = (args, opts = {}) => runtime.exec(args, { ...opts, timeoutMs: opts.timeoutMs ?? runTimeoutMs });
const ok = (result: ExecResult) => result.exit === 0;

function stripAnsi(value: string): string {
  return String(value ?? "").replace(/\x1b\]8;[^\x07]*\x07/g, "").replace(/\x1b\[[0-9;?]*[ -/]*[@-~]/g, "");
}

function snippet(value: string): string | undefined {
  const text = value?.trim();
  return text ? (text.length > 200 ? `${text.slice(0, 200)}…` : text) : undefined;
}

function resultDetail(label: string, result: ExecResult): string {
  const detail = snippet(result.err) ?? snippet(result.out);
  return `${label} failed (exit ${result.exit ?? -1})${detail ? ` — ${detail}` : ""}`;
}

// The keygen identity travels with the target so every probe names the
// machine key explicitly; in opt-out mode the target is unchanged.
function sshTarget(params: Opts): { ip?: string; user: string } & Opts {
  let ip = params.ip as string | undefined;
  if (params["provider-compute"] === "no-infra" && (!ip || ip === placeholderIp)) ip = params["no-infra-compute-ip"] as string;
  return {
    ip,
    user: String(params.user || params.sudoer || params["no-infra-compute-user"] || params["no-infra-compute-sudoer"] || "root"),
    ...(params["ssh-keygen"]
      ? { "ssh-keygen": params["ssh-keygen"], "ssh-private-key-path": params["ssh-private-key-path"] }
      : {}),
  };
}

function sshArgs(compute: { ip?: string; user: string } & Opts, remote: string[]): string[] {
  return ["ssh", "-o", "BatchMode=yes", "-o", "ConnectTimeout=5", "-o", "StrictHostKeyChecking=accept-new", ...identityArgs(compute), `${compute.user}@${compute.ip}`, ...remote];
}

async function computeStatus(runner: Runner, params: Opts, computeDetail?: string) {
  const external = params["provider-compute"] === "no-infra";
  const target = sshTarget(params);
  if (!target.ip || target.ip === placeholderIp) {
    return {
      ...target,
      status: external ? "unreachable" : "absent",
      detail: external ? "no host configured" : computeDetail ?? `no OpenTofu state in ${toolDir(params, "tofu-compute")}`,
    };
  }
  const result = await runner(sshArgs(target, ["true"]), { timeoutMs: sshProbeTimeoutMs });
  return { ...target, status: ok(result) ? "running" : "unreachable", detail: ok(result) ? "ssh ok" : resultDetail("ssh", result) };
}

export function providerSummary(params: Opts) {
  return { compute: params["provider-compute"], backend: params["provider-backend"], smtp: params["provider-smtp"], dns: params["provider-dns"] };
}

const hostStatusRe = /([A-Za-z0-9](?:[A-Za-z0-9.-]{0,253}[A-Za-z0-9])?\.[A-Za-z]{2,})(?:\s+\(([^)]*)\))?/;
export function parseOnceList(output: string): Array<{ host: string; status?: string }> {
  return stripAnsi(output).split(/\r?\n/).flatMap((line) => {
    const match = hostStatusRe.exec(line);
    return match ? [{ host: match[1]!, ...(match[2]?.trim() ? { status: match[2] } : {}) }] : [];
  });
}

export function imageRepositoryTag(image: unknown): { repository: string; tag: string; image: string } | undefined {
  const value = String(image ?? "").trim();
  if (!value || /^sha256:[A-Fa-f0-9]+$/.test(value)) return undefined;
  const withoutDigest = value.split("@", 1)[0]!;
  const slash = withoutDigest.lastIndexOf("/");
  const colon = withoutDigest.lastIndexOf(":");
  const tagged = colon > slash;
  const repository = tagged ? withoutDigest.slice(0, colon) : withoutDigest;
  const tag = tagged ? withoutDigest.slice(colon + 1) : "latest";
  return { repository, tag, image: `${repository}:${tag}` };
}

export function matchingRepoDigest(repository: string, digests: unknown[] = []): string | undefined {
  for (const entry of digests) {
    const [repo, digest] = String(entry).split("@", 2);
    if (repo === repository) return digest;
  }
}

function containerText(container: any): string {
  return JSON.stringify({ Name: container.Name, Config: container.Config, NetworkSettings: container.NetworkSettings }).toLowerCase();
}
// The host ONCE recorded on the container, from its `once` label. ONCE writes the
// application's desired state there as JSON, making this the authoritative
// container-to-host mapping; everything else is inference from names and labels.
function onceLabelHost(container: any): string | null {
  const raw = String(container?.Config?.Labels?.once ?? "").trim();
  if (!raw) return null;
  try {
    const host = JSON.parse(raw)?.host;
    return host ? String(host).trim().toLowerCase() || null : null;
  } catch {
    return null;
  }
}
// `example.com` sits inside `www.example.com`, so a plain substring lets the shorter host
// claim the longer one's container. Dots and alphanumerics block a match; `-` and `_` do
// not, because container names embed a dot-substituted host inside a longer name
// (`/once-www-example-com`).
function hostVariantRegExp(variant: string): RegExp {
  return new RegExp(`(?<![a-z0-9.])${variant.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}(?![a-z0-9.])`);
}
// The `once` label wins outright: a labelled container belongs to exactly the host it
// names and can never be claimed by another, so only unlabelled containers fall through
// to matching on names and third-party labels.
export function containerForHost(containers: any[], host: string): any {
  const lower = host.toLowerCase();
  const labelled = containers.find((container) => onceLabelHost(container) === lower);
  if (labelled) return labelled;
  const variants = [...new Set([lower, lower.replaceAll(".", "-"), lower.replaceAll(".", "_")])].map(hostVariantRegExp);
  return containers.find((container) => onceLabelHost(container) === null && variants.some((rx) => rx.test(containerText(container))));
}

async function registryDigest(runner: Runner, image: string, os: string, arch: string) {
  const args = ["skopeo", "inspect", "--no-tags", ...(os ? ["--override-os", os] : []), ...(arch ? ["--override-arch", arch] : []), `docker://${image}`];
  const result = await runner(args, { timeoutMs: registryTimeoutMs });
  if (!ok(result)) return { detail: resultDetail("skopeo inspect", result) };
  try { return { digest: JSON.parse(result.out).Digest as string }; }
  catch (error) { return { detail: `registry response was not valid JSON: ${error}` }; }
}

async function applicationReport(runner: Runner, containers: any[], images: any[], app: any) {
  const container = containerForHost(containers, app.host);
  const imageRef = container?.Config?.Image;
  const parsed = imageRepositoryTag(imageRef);
  const info = images.find((image) => image.Id === container?.Image || image.RepoTags?.includes(imageRef) || image.RepoTags?.includes(parsed?.image));
  const running = parsed ? matchingRepoDigest(parsed.repository, info?.RepoDigests) : undefined;
  const fallback = info?.Id ?? container?.Image;
  const registry = parsed ? await registryDigest(runner, parsed.image, info?.Os ?? "linux", info?.Architecture ?? "amd64") : {};
  return {
    host: app.host, status: container?.State?.Status ?? app.status ?? "unknown", image: parsed?.image ?? imageRef,
    version: parsed?.tag, digest: running ?? fallback, "digest-source": running ? "repo-digest" : fallback ? "image-id" : undefined,
    "registry-digest": (registry as any).digest,
    "new-version?": running && (registry as any).digest ? running !== (registry as any).digest : undefined,
    ...((registry as any).detail ? { "registry-detail": (registry as any).detail } : {}),
  };
}

async function remoteApplications(runner: Runner, compute: { ip?: string; user: string }) {
  const check = await runner(sshArgs(compute, ["command", "-v", "once", ">/dev/null", "2>&1", "||", "test", "-x", "/usr/local/bin/once", "||", "{", "echo", "once:", "command", "not", "found", ">&2", ";", "exit", "127", ";", "}"]));
  if (!ok(check)) return { ok: false, fatal: true, detail: resultDetail("once command check", check), applications: [] };
  const listed = await runner(sshArgs(compute, ["sudo", "-n", "once", "list"]));
  if (!ok(listed)) return { ok: false, fatal: listed.exit === 127 || /once: (command )?not found/i.test(`${listed.out}\n${listed.err}`), detail: resultDetail("once list", listed), applications: [] };
  const apps = parseOnceList(listed.out);
  if (!apps.length) return { ok: true, applications: [] };
  const ps = await runner(sshArgs(compute, ["sudo", "-n", "docker", "ps", "-q"]));
  if (!ok(ps)) return { ok: false, detail: resultDetail("docker ps", ps), applications: [] };
  const ids = ps.out.split(/\r?\n/).map((x) => x.trim()).filter(Boolean);
  if (!ids.length) return { ok: true, applications: apps.map((app) => ({ ...app, status: app.status ?? "unknown" })) };
  const inspected = await runner(sshArgs(compute, ["sudo", "-n", "docker", "inspect", "--type", "container", ...ids]));
  if (!ok(inspected)) return { ok: false, detail: resultDetail("docker inspect", inspected), applications: [] };
  const containers = JSON.parse(inspected.out || "[]");
  const imageIds = [...new Set(containers.flatMap((container: any) => [container.Image, container.Config?.Image]).filter(Boolean))] as string[];
  let images: any[] = [];
  if (imageIds.length) {
    const imageResult = await runner(sshArgs(compute, ["sudo", "-n", "docker", "image", "inspect", ...imageIds]));
    if (!ok(imageResult)) return { ok: false, detail: resultDetail("docker image inspect", imageResult), applications: [] };
    images = JSON.parse(imageResult.out || "[]");
  }
  return { ok: true, applications: await Promise.all(apps.map((app) => applicationReport(runner, containers, images, app))) };
}

async function tofuOutputParams(runner: Runner, opts: Opts, tool: string) {
  const result = await runner(["tofu", "output", "-json"], { cwd: toolDir(opts, tool), env: backendCredentialEnv(opts) });
  if (!ok(result)) return { params: {}, detail: resultDetail(`tofu output in ${toolDir(opts, tool)}`, result) };
  try { return { params: JSON.parse(result.out)?.params?.value ?? {} }; }
  catch (error) { return { params: {}, detail: `${tool} output was not valid JSON: ${error}` }; }
}

export async function describeReport(input: Opts, runner: Runner = run, resolve = true): Promise<any> {
  let opts = input;
  let detail: string | undefined;
  let computeDetail: string | undefined;
  if (resolve) {
    const loaded = await machine.load(opts);
    const compute = {params: loaded["once/compute-params"] as Record<string,unknown> ?? {}, detail: loaded["red/err"]};
    const smtp = await tofuOutputParams(runner, opts, "tofu-smtp");
    opts = { ...opts, ip: undefined, ...compute.params, ...smtp.params };
    computeDetail = compute.detail;
    detail = [compute.detail, smtp.detail].filter(Boolean).join("; ") || undefined;
  }
  const compute = await computeStatus(runner, opts, computeDetail);
  if (detail && compute.status !== "absent") compute.detail += `; ${detail}`;
  const appResult = compute.status === "running"
    ? await remoteApplications(runner, compute)
    : { ok: false, applications: [], detail: compute.status === "absent" ? "not checked because compute has not been created" : "not checked because compute is not reachable" };
  return {
    profile: opts.profile, providers: providerSummary(opts), compute,
    applications: appResult.applications ?? [],
    "applications-error": appResult.ok ? undefined : appResult.detail,
    "fatal-error?": Boolean((appResult as any).fatal),
  };
}

function present(value: unknown): string { return value === undefined || value === null || String(value).trim() === "" ? "unknown" : String(value); }
function updateLabel(value: unknown): string { return value === true ? "yes" : value === false ? "no" : "unknown"; }

export function printReport(report: any): void {
  console.log(`Profile: ${present(report.profile)}\n`);
  console.log(`Providers:\n  Compute: ${present(report.providers.compute)}\n  Backend: ${present(report.providers.backend)}\n  SMTP: ${present(report.providers.smtp)}\n  DNS: ${present(report.providers.dns)}\n`);
  console.log(`Compute:\n  IP: ${present(report.compute.ip)}\n  SSH user: ${present(report.compute.user)}\n  Status: ${present(report.compute.status)}${report.compute.detail ? ` (${report.compute.detail})` : ""}\n`);
  if (report["applications-error"]) console.log(`Applications: ${report["applications-error"]}.`);
  else if (!report.applications.length) console.log("Applications: none found.");
  else {
    console.log("Applications:");
    for (const app of report.applications) {
      console.log(`  - ${present(app.host)}\n    status: ${present(app.status)}\n    image: ${present(app.image)}\n    version: ${present(app.version)}\n    digest: ${present(app.digest)}${app["digest-source"] === "image-id" ? " (image id; digest comparison unknown)" : ""}\n    registry digest: ${present(app["registry-digest"])}\n    update available: ${updateLabel(app["new-version?"])}`);
      if (app["registry-detail"]) console.log(`    registry check: ${app["registry-detail"]}`);
    }
  }
}

export async function describe(opts: Opts): Promise<Opts> {
  const result = await describeReport(opts);
  printReport(result);
  const computeError = result.compute.status === "running" ? undefined : `compute is ${result.compute.status}${result.compute.detail ? ` — ${result.compute.detail}` : ""}`;
  if (result["fatal-error?"]) return { ...opts, "once.describe/result": result, "red/exit": 1, "red/err": result["applications-error"] ?? "describe failed" };
  return { ...opts, "once.describe/result": result, ...(computeError ? { "red/exit": 1, "red/err": computeError } : { "red/exit": 0 }) };
}

export async function describeFile(path: string): Promise<Opts> {
  try {
    if (!existsSync(path)) return { "red/exit": 2, "red/err": `desired state file not found: ${path}` };
    // Describe reads the live host, so it needs the keygen identity the way
    // create does — real event semantics, opt-out untouched.
    return describe(withMachineKey(readPars({
      ...((Bun.YAML.parse(readFileSync(path, "utf8")) ?? {}) as Opts),
      "red/state-file": resolve(path),
    }), true));
  } catch (error) {
    return { "red/exit": 2, "red/err": error instanceof Error ? error.message : String(error) };
  }
}

File: blue/src/package_once_blue/describe.py
from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any, Awaitable, Callable

from blue.cli import load_yaml
from blue.runtime import ExecResult, runtime

from blue.cli import read_pars

from .ssh import identity_args, with_machine_key
from . import machine
from .tools import backend_credential_env, tool_dir

Runner = Callable[..., Awaitable[ExecResult]]
_PLACEHOLDER_IP = "192.168.0.1"
_ANSI_RE = re.compile(r"\x1b\]8;[^\x07]*\x07|\x1b\[[0-9;?]*[ -/]*[@-~]")
_HOST_RE = re.compile(r"([A-Za-z0-9](?:[A-Za-z0-9.-]{0,253}[A-Za-z0-9])?\.[A-Za-z]{2,})(?:\s+\(([^)]*)\))?")


async def _run(args: list[str], *, cwd: str | None = None, env: dict | None = None, timeout_ms: int = 30_000) -> ExecResult:
    return await runtime.exec(args, cwd=cwd, env=env, timeout_ms=timeout_ms)


def _detail(label: str, result: ExecResult) -> str:
    text = (result.err or result.out or "").strip()
    if len(text) > 200:
        text = text[:200] + "…"
    return f"{label} failed (exit {result.exit})" + (f" — {text}" if text else "")


def _target(params: dict) -> dict:
    ip = params.get("ip")
    if params.get("provider-compute") == "no-infra" and (not ip or ip == _PLACEHOLDER_IP):
        ip = params.get("no-infra-compute-ip")
    # The keygen identity travels with the target so every probe names the
    # machine key explicitly; in opt-out mode the target is unchanged.
    identity = {"ssh-keygen": params.get("ssh-keygen"), "ssh-private-key-path": params.get("ssh-private-key-path")} if params.get("ssh-keygen") else {}
    return {"ip": ip, "user": params.get("user") or params.get("sudoer") or params.get("no-infra-compute-user") or params.get("no-infra-compute-sudoer") or "root", **identity}


def _ssh(target: dict, remote: list[str]) -> list[str]:
    return ["ssh", "-o", "BatchMode=yes", "-o", "ConnectTimeout=5", "-o", "StrictHostKeyChecking=accept-new", *identity_args(target), f"{target['user']}@{target['ip']}", *remote]


async def _compute_status(runner: Runner, params: dict, state_detail: str | None = None) -> dict:
    target = _target(params)
    external = params.get("provider-compute") == "no-infra"
    if not target.get("ip") or target["ip"] == _PLACEHOLDER_IP:
        return {**target, "status": "unreachable" if external else "absent", "detail": "no host configured" if external else state_detail or f"no OpenTofu state in {tool_dir(params, 'tofu-compute')}"}
    result = await runner(_ssh(target, ["true"]), timeout_ms=10_000)
    return {**target, "status": "running" if result.exit == 0 else "unreachable", "detail": "ssh ok" if result.exit == 0 else _detail("ssh", result)}


def provider_summary(params: dict) -> dict:
    return {"compute": params.get("provider-compute"), "backend": params.get("provider-backend"), "smtp": params.get("provider-smtp"), "dns": params.get("provider-dns")}


def parse_once_list(output: str) -> list[dict]:
    result = []
    for line in _ANSI_RE.sub("", output or "").splitlines():
        if match := _HOST_RE.search(line):
            result.append({"host": match.group(1), **({"status": match.group(2)} if match.group(2) else {})})
    return result


def image_repository_tag(image: object) -> dict | None:
    value = str(image or "").strip()
    if not value or re.fullmatch(r"sha256:[A-Fa-f0-9]+", value):
        return None
    value = value.split("@", 1)[0]
    slash, colon = value.rfind("/"), value.rfind(":")
    repository, tag = (value[:colon], value[colon + 1 :]) if colon > slash else (value, "latest")
    return {"repository": repository, "tag": tag, "image": f"{repository}:{tag}"}


def matching_repo_digest(repository: str, digests: list | None) -> str | None:
    for item in digests or []:
        repo, _, digest = str(item).partition("@")
        if repo == repository:
            return digest
    return None


def _once_label_host(container: dict) -> str | None:
    """The host ONCE recorded on the container, from its `once` label. ONCE writes the
    application's desired state there as JSON, making this the authoritative
    container-to-host mapping; everything else is inference from names and labels."""
    raw = str(((container.get("Config") or {}).get("Labels") or {}).get("once") or "").strip()
    if not raw:
        return None
    try:
        host = json.loads(raw).get("host")
    except (ValueError, AttributeError):
        return None
    return str(host).strip().lower() or None if host else None


def _host_variant_re(variant: str) -> re.Pattern:
    """`example.com` sits inside `www.example.com`, so a plain substring lets the shorter
    host claim the longer one's container. Dots and alphanumerics block a match; `-` and
    `_` do not, because container names embed a dot-substituted host inside a longer
    name (`/once-www-example-com`)."""
    return re.compile(rf"(?<![a-z0-9.]){re.escape(variant)}(?![a-z0-9.])")


def _container_text(container: dict) -> str:
    return json.dumps({k: container.get(k) for k in ["Name", "Config", "NetworkSettings"]}).lower()


def _container_for_host(containers: list[dict], host: str) -> dict | None:
    """The `once` label wins outright: a labelled container belongs to exactly the host it
    names and can never be claimed by another, so only unlabelled containers fall
    through to matching on names and third-party labels."""
    host = host.lower()
    if labelled := next((c for c in containers if _once_label_host(c) == host), None):
        return labelled
    variants = [_host_variant_re(v) for v in dict.fromkeys([host, host.replace(".", "-"), host.replace(".", "_")])]
    return next((c for c in containers if _once_label_host(c) is None and any(rx.search(_container_text(c)) for rx in variants)), None)


async def _application_report(runner: Runner, containers: list, images: list, app: dict) -> dict:
    container = _container_for_host(containers, app["host"])
    image_ref = (container or {}).get("Config", {}).get("Image")
    parsed = image_repository_tag(image_ref)
    image = next((info for info in images if info.get("Id") == (container or {}).get("Image") or image_ref in info.get("RepoTags", []) or (parsed and parsed["image"] in info.get("RepoTags", []))), None)
    running = matching_repo_digest(parsed["repository"], (image or {}).get("RepoDigests")) if parsed else None
    fallback = (image or {}).get("Id") or (container or {}).get("Image")
    registry_digest = registry_detail = None
    if parsed:
        result = await runner(["skopeo", "inspect", "--no-tags", "--override-os", (image or {}).get("Os", "linux"), "--override-arch", (image or {}).get("Architecture", "amd64"), f"docker://{parsed['image']}"], timeout_ms=30_000)
        if result.exit == 0:
            try:
                registry_digest = json.loads(result.out).get("Digest")
            except Exception as error:
                registry_detail = f"registry response was not valid JSON: {error}"
        else:
            registry_detail = _detail("skopeo inspect", result)
    return {"host": app["host"], "status": (container or {}).get("State", {}).get("Status") or app.get("status") or "unknown", "image": (parsed or {}).get("image") or image_ref, "version": (parsed or {}).get("tag"), "digest": running or fallback, "digest-source": "repo-digest" if running else "image-id" if fallback else None, "registry-digest": registry_digest, "new-version?": running != registry_digest if running and registry_digest else None, **({"registry-detail": registry_detail} if registry_detail else {})}


async def _remote_apps(runner: Runner, compute: dict) -> dict:
    check = await runner(_ssh(compute, ["command", "-v", "once", ">/dev/null", "2>&1", "||", "test", "-x", "/usr/local/bin/once", "||", "exit", "127"]))
    if check.exit != 0:
        return {"ok": False, "fatal": True, "detail": _detail("once command check", check), "applications": []}
    listed = await runner(_ssh(compute, ["sudo", "-n", "once", "list"]))
    if listed.exit != 0:
        return {"ok": False, "fatal": listed.exit == 127 or "once: not found" in (listed.out + listed.err).lower(), "detail": _detail("once list", listed), "applications": []}
    apps = parse_once_list(listed.out)
    if not apps:
        return {"ok": True, "applications": []}
    ps = await runner(_ssh(compute, ["sudo", "-n", "docker", "ps", "-q"]))
    if ps.exit != 0:
        return {"ok": False, "detail": _detail("docker ps", ps), "applications": []}
    ids = [line.strip() for line in ps.out.splitlines() if line.strip()]
    if not ids:
        return {"ok": True, "applications": apps}
    inspected = await runner(_ssh(compute, ["sudo", "-n", "docker", "inspect", "--type", "container", *ids]))
    if inspected.exit != 0:
        return {"ok": False, "detail": _detail("docker inspect", inspected), "applications": []}
    containers = json.loads(inspected.out or "[]")
    image_ids = list(dict.fromkeys(x for container in containers for x in [container.get("Image"), container.get("Config", {}).get("Image")] if x))
    images = []
    if image_ids:
        result = await runner(_ssh(compute, ["sudo", "-n", "docker", "image", "inspect", *image_ids]))
        if result.exit != 0:
            return {"ok": False, "detail": _detail("docker image inspect", result), "applications": []}
        images = json.loads(result.out or "[]")
    return {"ok": True, "applications": [await _application_report(runner, containers, images, app) for app in apps]}


async def _tofu_params(runner: Runner, opts: dict, tool: str) -> dict:
    result = await runner(["tofu", "output", "-json"], cwd=tool_dir(opts, tool), env=backend_credential_env(opts))
    if result.exit != 0:
        return {"params": {}, "detail": _detail(f"tofu output in {tool_dir(opts, tool)}", result)}
    try:
        return {"params": json.loads(result.out).get("params", {}).get("value", {})}
    except Exception as error:
        return {"params": {}, "detail": f"{tool} output was not valid JSON: {error}"}


async def describe_report(input: dict, runner: Runner = _run, resolve: bool = True) -> dict:
    opts, detail, compute_detail = input, None, None
    if resolve:
        loaded = await machine.load(opts)
        compute = {"params": loaded.get("once/compute-params", {}), "detail": loaded.get("blue/err")}
        smtp = await _tofu_params(runner, opts, "tofu-smtp")
        opts = {**{k: v for k, v in opts.items() if k != "ip"}, **compute["params"], **smtp["params"]}
        compute_detail = compute.get("detail")
        detail = "; ".join(x for x in [compute.get("detail"), smtp.get("detail")] if x) or None
    compute = await _compute_status(runner, opts, compute_detail)
    if detail and compute["status"] != "absent":
        compute["detail"] += f"; {detail}"
    apps = await _remote_apps(runner, compute) if compute["status"] == "running" else {"ok": False, "applications": [], "detail": "not checked because compute has not been created" if compute["status"] == "absent" else "not checked because compute is not reachable"}
    return {"profile": opts.get("profile"), "providers": provider_summary(opts), "compute": compute, "applications": apps.get("applications", []), "applications-error": None if apps.get("ok") else apps.get("detail"), "fatal-error?": bool(apps.get("fatal"))}


def _present(value: Any) -> str:
    return "unknown" if value is None or not str(value).strip() else str(value)


def print_report(report: dict) -> None:
    providers, compute = report["providers"], report["compute"]
    print(f"Profile: {_present(report.get('profile'))}\n\nProviders:\n  Compute: {_present(providers.get('compute'))}\n  Backend: {_present(providers.get('backend'))}\n  SMTP: {_present(providers.get('smtp'))}\n  DNS: {_present(providers.get('dns'))}\n")
    print(f"Compute:\n  IP: {_present(compute.get('ip'))}\n  SSH user: {_present(compute.get('user'))}\n  Status: {_present(compute.get('status'))}" + (f" ({compute['detail']})" if compute.get("detail") else "") + "\n")
    if report.get("applications-error"):
        print(f"Applications: {report['applications-error']}.")
    elif not report["applications"]:
        print("Applications: none found.")
    else:
        print("Applications:")
        for app in report["applications"]:
            update = "yes" if app.get("new-version?") is True else "no" if app.get("new-version?") is False else "unknown"
            print(f"  - {_present(app.get('host'))}\n    status: {_present(app.get('status'))}\n    image: {_present(app.get('image'))}\n    version: {_present(app.get('version'))}\n    digest: {_present(app.get('digest'))}\n    registry digest: {_present(app.get('registry-digest'))}\n    update available: {update}")


async def describe(opts: dict) -> dict:
    report = await describe_report(opts)
    print_report(report)
    if report.get("fatal-error?"):
        return {**opts, "once.describe/result": report, "blue/exit": 1, "blue/err": report.get("applications-error") or "describe failed"}
    if report["compute"]["status"] != "running":
        compute = report["compute"]
        return {**opts, "once.describe/result": report, "blue/exit": 1, "blue/err": f"compute is {compute['status']}" + (f" — {compute['detail']}" if compute.get("detail") else "")}
    return {**opts, "once.describe/result": report, "blue/exit": 0}


async def describe_file(path: str) -> dict:
    try:
        file = Path(path)
        if not file.exists():
            return {"blue/exit": 2, "blue/err": f"desired state file not found: {path}"}
        # Describe reads the live host, so it needs the keygen identity the
        # way create does — real event semantics, opt-out untouched.
        return await describe(with_machine_key(read_pars({
            **load_yaml(file.read_text()),
            "blue/state-file": str(file.resolve()),
        }), True))
    except Exception as error:
        return {"blue/exit": 2, "blue/err": str(error) or type(error).__name__}
